File: java/liskov-substitution/LSPCorrectDemonstration.java
import java.util.*;

/**
 * CORRECT IMPLEMENTATION - FOLLOWS LISKOV SUBSTITUTION PRINCIPLE
 * 
 * This demonstrates proper inheritance where subtypes can be substituted
 * for their base types without breaking the program's correctness.
 * 
 * Key LSP principle: Objects of a superclass should be replaceable with
 * objects of a subclass without altering the correctness of the program.
 */
public class LSPCorrectDemonstration {
    
    // Abstract base class defining the contract
    static abstract class DataStructure<T> {
        protected List<T> elements = new ArrayList<>();
        
        public abstract void add(T element);
        public abstract T remove();
        public abstract T peek();
        public abstract boolean isEmpty();
        public abstract int size();
        
        // This method should work with any proper implementation
        public void transferTo(DataStructure<T> other, int count) {
            for (int i = 0; i < count && !this.isEmpty(); i++) {
                T element = this.remove();
                other.add(element);
            }
        }
        
        @Override
        public String toString() {
            return getClass().getSimpleName() + ": " + elements.toString();
        }
    }
    
    // Stack implementation that follows LSP
    static class ProperStack<T> extends DataStructure<T> {
        @Override
        public void add(T element) {
            elements.add(element); // Add to end (top of stack)
        }
        
        @Override
        public T remove() {
            if (isEmpty()) throw new EmptyStackException();
            return elements.remove(elements.size() - 1); // Remove from end (LIFO)
        }
        
        @Override
        public T peek() {
            if (isEmpty()) throw new EmptyStackException();
            return elements.get(elements.size() - 1); // Peek at end
        }
        
        @Override
        public boolean isEmpty() {
            return elements.isEmpty();
        }
        
        @Override
        public int size() {
            return elements.size();
        }
    }
    
    // Queue implementation that follows LSP
    static class ProperQueue<T> extends DataStructure<T> {
        @Override
        public void add(T element) {
            elements.add(element); // Add to end
        }
        
        @Override
        public T remove() {
            if (isEmpty()) throw new NoSuchElementException("Queue is empty");
            return elements.remove(0); // Remove from beginning (FIFO)
        }
        
        @Override
        public T peek() {
            if (isEmpty()) throw new NoSuchElementException("Queue is empty");
            return elements.get(0); // Peek at beginning
        }
        
        @Override
        public boolean isEmpty() {
            return elements.isEmpty();
        }
        
        @Override
        public int size() {
            return elements.size();
        }
    }
    
    // Priority Queue implementation that follows LSP
    static class ProperPriorityQueue<T extends Comparable<T>> extends DataStructure<T> {
        @Override
        public void add(T element) {
            elements.add(element);
            Collections.sort(elements); // Keep sorted for priority
        }
        
        @Override
        public T remove() {
            if (isEmpty()) throw new NoSuchElementException("Priority queue is empty");
            return elements.remove(elements.size() - 1); // Remove highest priority (last after sorting)
        }
        
        @Override
        public T peek() {
            if (isEmpty()) throw new NoSuchElementException("Priority queue is empty");
            return elements.get(elements.size() - 1); // Peek at highest priority
        }
        
        @Override
        public boolean isEmpty() {
            return elements.isEmpty();
        }
        
        @Override
        public int size() {
            return elements.size();
        }
    }
    
    // Client code that works with any DataStructure implementation
    public static void processDataStructure(DataStructure<Integer> ds, String name) {
        System.out.println("--- Processing " + name + " ---");
        
        // Add elements
        ds.add(10);
        ds.add(20);
        ds.add(30);
        System.out.println("After adding 10, 20, 30: " + ds);
        
        // Peek at next element
        System.out.println("Next element (peek): " + ds.peek());
        System.out.println("Size: " + ds.size());
        
        // Remove elements
        System.out.println("Removing: " + ds.remove());
        System.out.println("After removal: " + ds);
        
        // Check if operations maintain the expected contract
        System.out.println("Is empty: " + ds.isEmpty());
        System.out.println("Size after removal: " + ds.size());
        System.out.println();
    }
    
    public static void demonstrateTransfer() {
        System.out.println("=== TRANSFER OPERATION DEMONSTRATION ===");
        
        DataStructure<Integer> stack = new ProperStack<>();
        DataStructure<Integer> queue = new ProperQueue<>();
        
        // Populate stack
        stack.add(1);
        stack.add(2);
        stack.add(3);
        
        System.out.println("Before transfer:");
        System.out.println("Stack: " + stack);
        System.out.println("Queue: " + queue);
        
        // Transfer 2 elements from stack to queue
        stack.transferTo(queue, 2);
        
        System.out.println("After transferring 2 elements from stack to queue:");
        System.out.println("Stack: " + stack);
        System.out.println("Queue: " + queue);
        
        // This works because both implementations properly follow the contract
        System.out.println();
    }
    
    public static void main(String[] args) {
        System.out.println("=== LSP CORRECT DEMONSTRATION ===");
        System.out.println("All implementations can be substituted without breaking correctness");
        System.out.println();
        
        // Test all implementations with the same client code
        DataStructure<Integer> stack = new ProperStack<>();
        DataStructure<Integer> queue = new ProperQueue<>();
        DataStructure<Integer> priorityQueue = new ProperPriorityQueue<>();
        
        // All these calls work correctly because LSP is followed
        processDataStructure(stack, "Stack (LIFO)");
        processDataStructure(queue, "Queue (FIFO)");
        processDataStructure(priorityQueue, "Priority Queue");
        
        // Demonstrate polymorphic behavior
        demonstrateTransfer();
        
        // Show that behavior is predictable and consistent
        System.out.println("=== POLYMORPHIC ARRAY DEMONSTRATION ===");
        DataStructure<Integer>[] structures = {
            new ProperStack<>(),
            new ProperQueue<>(),
            new ProperPriorityQueue<>()
        };
        
        for (DataStructure<Integer> ds : structures) {
            ds.add(5);
            ds.add(1);
            ds.add(3);
            System.out.println(ds.getClass().getSimpleName() + " peek: " + ds.peek());
        }
        
        System.out.println();
        System.out.println("=== WHY THIS FOLLOWS LSP ===");
        System.out.println("1. All subtypes honor the base class contract");
        System.out.println("2. Method preconditions are not strengthened in subtypes");
        System.out.println("3. Method postconditions are not weakened in subtypes");
        System.out.println("4. Base class invariants are preserved in subtypes");
        System.out.println("5. Client code works correctly with any implementation");
        System.out.println("6. Behavioral substitutability is maintained");
    }
}

File: python/liskov-substitution/lsp_correct_demonstration.py
"""
CORRECT IMPLEMENTATION - FOLLOWS LISKOV SUBSTITUTION PRINCIPLE

This demonstrates proper inheritance where subtypes can be substituted
for their base types without breaking the program's correctness.

In Python, LSP is enforced through proper inheritance design and
behavioral contracts, even though the language is dynamically typed.
"""

from abc import ABC, abstractmethod
from typing import List, Optional
import random


class DataStructure(ABC):
    """Abstract base class defining the contract for data structures"""
    
    @abstractmethod
    def add(self, element: int) -> None:
        """Add an element to the data structure"""
        pass
    
    @abstractmethod
    def remove(self) -> int:
        """Remove and return an element from the data structure"""
        pass
    
    @abstractmethod
    def peek(self) -> int:
        """Look at the next element without removing it"""
        pass
    
    @abstractmethod
    def size(self) -> int:
        """Return the number of elements"""
        pass
    
    @abstractmethod
    def is_empty(self) -> bool:
        """Check if the data structure is empty"""
        pass


class Stack(DataStructure):
    """Stack implementation (LIFO - Last In, First Out)"""
    
    def __init__(self):
        self._items: List[int] = []
    
    def add(self, element: int) -> None:
        self._items.append(element)
    
    def remove(self) -> int:
        if not self._items:
            raise IndexError("Stack is empty")
        return self._items.pop()  # Remove from end (LIFO)
    
    def peek(self) -> int:
        if not self._items:
            raise IndexError("Stack is empty")
        return self._items[-1]  # Peek at end
    
    def size(self) -> int:
        return len(self._items)
    
    def is_empty(self) -> bool:
        return len(self._items) == 0
    
    def __str__(self) -> str:
        return f"Stack{self._items}"


class Queue(DataStructure):
    """Queue implementation (FIFO - First In, First Out)"""
    
    def __init__(self):
        self._items: List[int] = []
    
    def add(self, element: int) -> None:
        self._items.append(element)  # Add to end
    
    def remove(self) -> int:
        if not self._items:
            raise IndexError("Queue is empty")
        return self._items.pop(0)  # Remove from beginning (FIFO)
    
    def peek(self) -> int:
        if not self._items:
            raise IndexError("Queue is empty")
        return self._items[0]  # Peek at beginning
    
    def size(self) -> int:
        return len(self._items)
    
    def is_empty(self) -> bool:
        return len(self._items) == 0
    
    def __str__(self) -> str:
        return f"Queue{self._items}"


class PriorityQueue(DataStructure):
    """Priority Queue implementation (highest value first)"""
    
    def __init__(self):
        self._items: List[int] = []
    
    def add(self, element: int) -> None:
        self._items.append(element)
        self._items.sort()  # Keep sorted for priority
    
    def remove(self) -> int:
        if not self._items:
            raise IndexError("Priority queue is empty")
        return self._items.pop()  # Remove highest (last after sorting)
    
    def peek(self) -> int:
        if not self._items:
            raise IndexError("Priority queue is empty")
        return self._items[-1]  # Peek at highest
    
    def size(self) -> int:
        return len(self._items)
    
    def is_empty(self) -> bool:
        return len(self._items) == 0
    
    def __str__(self) -> str:
        return f"PriorityQueue{self._items}"


class Deque(DataStructure):
    """Double-ended queue that can add/remove from both ends"""
    
    def __init__(self, mode='back'):
        """Mode can be 'front' or 'back' to specify default removal end"""
        self._items: List[int] = []
        self._mode = mode
    
    def add(self, element: int) -> None:
        self._items.append(element)  # Always add to back
    
    def remove(self) -> int:
        if not self._items:
            raise IndexError("Deque is empty")
        
        if self._mode == 'front':
            return self._items.pop(0)  # Remove from front
        else:
            return self._items.pop()   # Remove from back
    
    def peek(self) -> int:
        if not self._items:
            raise IndexError("Deque is empty")
        
        if self._mode == 'front':
            return self._items[0]   # Peek at front
        else:
            return self._items[-1]  # Peek at back
    
    def size(self) -> int:
        return len(self._items)
    
    def is_empty(self) -> bool:
        return len(self._items) == 0
    
    def __str__(self) -> str:
        return f"Deque({self._mode}){self._items}"


# Client functions that work with any DataStructure implementation
def process_data_structure(ds: DataStructure, name: str) -> None:
    """Process any data structure following the contract"""
    print(f"--- Processing {name} ---")
    
    # Add elements
    ds.add(10)
    ds.add(20)
    ds.add(30)
    print(f"After adding 10, 20, 30 - Size: {ds.size()}")
    
    # Peek at next element
    try:
        next_element = ds.peek()
        print(f"Next element (peek): {next_element}")
    except IndexError as e:
        print(f"Peek error: {e}")
    
    # Remove elements
    try:
        element = ds.remove()
        print(f"Removed: {element}")
    except IndexError as e:
        print(f"Remove error: {e}")
    
    print(f"Size after removal: {ds.size()}")
    print(f"Is empty: {ds.is_empty()}")
    print()


def transfer_elements(source: DataStructure, target: DataStructure, count: int) -> None:
    """Transfer elements from source to target"""
    for _ in range(count):
        if source.is_empty():
            break
        try:
            element = source.remove()
            target.add(element)
        except IndexError:
            break


def count_elements(ds: DataStructure) -> int:
    """Count elements by removing them (destructive)"""
    count = 0
    while not ds.is_empty():
        try:
            ds.remove()
            count += 1
        except IndexError:
            break
    return count


def sum_elements(ds: DataStructure) -> int:
    """Sum all elements by removing them (destructive)"""
    total = 0
    while not ds.is_empty():
        try:
            element = ds.remove()
            total += element
        except IndexError:
            break
    return total


def demonstrate_polymorphism():
    """Show how different implementations can be used polymorphically"""
    print("=== POLYMORPHIC BEHAVIOR DEMONSTRATION ===")
    
    # Create different data structures
    structures = [
        Stack(),
        Queue(),
        PriorityQueue(),
        Deque('back'),
        Deque('front')
    ]
    
    names = [
        "Stack",
        "Queue", 
        "PriorityQueue",
        "Deque(back)",
        "Deque(front)"
    ]
    
    # All can be treated the same way
    for ds, name in zip(structures, names):
        ds.add(5)
        ds.add(1)
        ds.add(3)
        
        try:
            next_element = ds.peek()
            print(f"{name} peek: {next_element}")
        except IndexError:
            print(f"{name} peek: empty")
    
    print()


def demonstrate_transfer():
    """Demonstrate transfer between different data structure types"""
    print("=== TRANSFER OPERATION DEMONSTRATION ===")
    
    source = Stack()
    target = Queue()
    
    # Populate source
    for i in [1, 2, 3, 4, 5]:
        source.add(i)
    
    print(f"Before transfer - Source: {source}, Target: {target}")
    
    # Transfer 3 elements
    transfer_elements(source, target, 3)
    
    print(f"After transfer - Source: {source}, Target: {target}")
    print()


def demonstrate_processors():
    """Demonstrate functions that process any data structure"""
    print("=== PROCESSOR DEMONSTRATION ===")
    
    # Create and populate different structures
    stack = Stack()
    queue = Queue()
    
    for i in [10, 20, 30, 40, 50]:
        stack.add(i)
        queue.add(i)
    
    # Process with different functions
    print(f"Stack before counting: {stack}")
    count = count_elements(Stack())  # Use fresh stack
    for i in [10, 20, 30, 40, 50]:
        Stack().add(i)  # Can't modify original, so create new one
    
    # Better approach: clone the structures
    stack_copy = Stack()
    queue_copy = Queue()
    
    for i in [10, 20, 30, 40, 50]:
        stack_copy.add(i)
        queue_copy.add(i)
    
    stack_count = count_elements(stack_copy)
    queue_sum = sum_elements(queue_copy)
    
    print(f"Stack element count: {stack_count}")
    print(f"Queue element sum: {queue_sum}")
    print()


def main():
    print("=== LSP CORRECT DEMONSTRATION ===")
    print("All implementations can be substituted without breaking correctness")
    print()
    
    # Test all implementations with the same client code
    stack = Stack()
    queue = Queue()
    priority_queue = PriorityQueue()
    deque_back = Deque('back')
    deque_front = Deque('front')
    
    # All these calls work correctly because LSP is followed
    process_data_structure(stack, "Stack (LIFO)")
    process_data_structure(queue, "Queue (FIFO)")
    process_data_structure(priority_queue, "Priority Queue (Highest First)")
    process_data_structure(deque_back, "Deque (Back Mode)")
    process_data_structure(deque_front, "Deque (Front Mode)")
    
    # Demonstrate polymorphic behavior
    demonstrate_polymorphism()
    
    # Demonstrate transfer between different types
    demonstrate_transfer()
    
    # Demonstrate processors working with any data structure
    demonstrate_processors()
    
    print("=== WHY THIS FOLLOWS LSP ===")
    print("1. All implementations honor the DataStructure abstract contract")
    print("2. Client code works correctly with any implementation")
    print("3. Behavioral substitutability is maintained")
    print("4. Exception handling is consistent across implementations")
    print("5. No unexpected side effects or behavior changes")
    print("6. Preconditions are not strengthened in subclasses")
    print("7. Postconditions are not weakened in subclasses")


if __name__ == "__main__":
    main()